packages/ember-metal/lib/property_set.js

```diff
@@ -100,16 +100,13 @@ if (MANDATORY_SETTER) {
 
 function setPath(root, path, value, tolerant) {
   // get the last part of the path
-  let keyName = path.slice(path.lastIndexOf('.') + 1);
+  let parts = path.split('.');
+  let keyName = parts.pop();
 
   // get the first part of the part
-  path = (path === keyName) ? keyName : path.slice(0, path.length - (keyName.length + 1));
+  path = (path === keyName) ? keyName : parts.join('.');
 
-  // unless the path is this, look up the first part to
-  // get the root
-  if (path !== 'this') {
```

Comment:
there is already assertion to not allow that https://github.com/bekzod/ember.js/blob/57f47143694d0db1197bfec0a582a40defbb0ca9/packages/ember-metal/lib/property_set.js#L43